File: src/main/java/no/nav/sosialhjelp/soknad/web/service/SaksoversiktMetadataService.java
package no.nav.sosialhjelp.soknad.web.service;

import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.EttersendingsSoknad;
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.Hoveddokument;
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.InnsendtSoknad;
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.PabegyntSoknad;
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.Part;
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.Vedlegg;
import no.nav.sosialhjelp.soknad.business.db.repositories.soknadmetadata.SoknadMetadataRepository;
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata;
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata.VedleggMetadataListe;
import no.nav.sosialhjelp.soknad.business.service.soknadservice.EttersendingService;
import no.nav.sosialhjelp.soknad.tekster.NavMessageSource;
import org.springframework.stereotype.Service;

import java.time.Clock;
import java.time.LocalDateTime;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Properties;

import static java.util.stream.Collectors.toList;
import static no.nav.sosialhjelp.soknad.business.service.soknadservice.EttersendingService.ETTERSENDELSE_FRIST_DAGER;
import static no.nav.sosialhjelp.soknad.business.util.JsonVedleggUtils.isVedleggskravAnnet;
import static no.nav.sosialhjelp.soknad.domain.Vedleggstatus.LastetOpp;
import static no.nav.sosialhjelp.soknad.domain.Vedleggstatus.VedleggKreves;
import static no.nav.sosialhjelp.soknad.domain.model.kravdialoginformasjon.SoknadType.SEND_SOKNAD_KOMMUNAL;

@Service
public class SaksoversiktMetadataService {

    private final SoknadMetadataRepository soknadMetadataRepository;
    private final EttersendingService ettersendingService;
    private final NavMessageSource navMessageSource;
    private final Clock clock;

    public SaksoversiktMetadataService(
            SoknadMetadataRepository soknadMetadataRepository,
            EttersendingService ettersendingService,
            NavMessageSource navMessageSource,
            Clock clock
    ) {
        this.soknadMetadataRepository = soknadMetadataRepository;
        this.ettersendingService = ettersendingService;
        this.navMessageSource = navMessageSource;
        this.clock = clock;
    }

    public List<InnsendtSoknad> hentInnsendteSoknaderForFnr(String fnr) {
        Properties bundle = getBundle();

        List<SoknadMetadata> soknader = soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(fnr);

        return soknader.stream().map(soknad ->
                new InnsendtSoknad()
                        .withAvsender(new Part()
                                .withType(Part.Type.BRUKER)
                                .withVisningsNavn(bundle.getProperty("saksoversikt.mottaker.deg")))
                        .withMottaker(new Part()
                                .withType(Part.Type.NAV)
                                .withVisningsNavn(bundle.getProperty("saksoversikt.mottaker.nav")))
                        .withBehandlingsId(soknad.behandlingsId)
                        .withInnsendtDato(tilDate(soknad.innsendtDato))
                        .withHoveddokument(new Hoveddokument()
                                .withTittel(soknad.type.equals(SEND_SOKNAD_KOMMUNAL) ? bundle.getProperty("saksoversikt.soknadsnavn") : bundle.getProperty("saksoversikt.soknadsnavn.ettersending")))
                        .withVedlegg(tilInnsendteVedlegg(soknad.vedlegg, bundle))
                        .withTema("KOM")
                        .withTemanavn(bundle.getProperty("saksoversikt.temanavn"))
                        .withLenke(lagEttersendelseLenke(soknad.behandlingsId)))
                .collect(toList());
    }

    public List<PabegyntSoknad> hentPabegynteSoknaderForBruker(String fnr) {
        List<SoknadMetadata> soknader = soknadMetadataRepository.hentPabegynteSoknaderForBruker(fnr);

        return soknader.stream().map(soknad ->
                new PabegyntSoknad()
                        .withBehandlingsId(soknad.behandlingsId)
                        .withTittel("Søknad om økonomisk sosialhjelp")
                        .withSisteEndring(tilDate(soknad.sistEndretDato))
                        .withLenke(lagFortsettSoknadLenke(soknad.behandlingsId))
        ).collect(toList());
    }

    public List<EttersendingsSoknad> hentSoknaderBrukerKanEttersendePa(String fnr) {
        Properties bundle = getBundle();
        LocalDateTime ettersendelseFrist = LocalDateTime.now(clock)
                .minusDays(ETTERSENDELSE_FRIST_DAGER);
        DateTimeFormatter datoFormatter = DateTimeFormatter.ofPattern("d. MMMM yyyy");

        List<SoknadMetadata> soknader = soknadMetadataRepository.hentInnsendteSoknaderForBrukerEtterTidspunkt(fnr, ettersendelseFrist);

        return soknader.stream().map(soknad ->
            new EttersendingsSoknad()
                .withBehandlingsId(soknad.behandlingsId)
                .withTittel(bundle.getProperty("saksoversikt.soknadsnavn") + " (" + soknad.innsendtDato.format(datoFormatter) +  ")")
                .withLenke(lagEttersendelseLenke(soknad.behandlingsId))
                .withVedlegg(finnManglendeVedlegg(soknad, bundle))
        ).collect(toList());
    }

    private List<Vedlegg> finnManglendeVedlegg(SoknadMetadata soknad, Properties bundle) {
        SoknadMetadata nyesteSoknad = ettersendingService.hentNyesteSoknadIKjede(soknad);

        return nyesteSoknad.vedlegg.vedleggListe.stream()
                .filter(v -> v.status.er(VedleggKreves))
                .filter(v -> !(isVedleggskravAnnet(v)))
                .map(v -> "vedlegg." + v.skjema + "." + v.tillegg + ".tittel")
                .distinct()
                .map(bundle::getProperty)
                .map(navn -> new Vedlegg().withTittel(navn))
                .collect(toList());
    }

    private List<Vedlegg> tilInnsendteVedlegg(VedleggMetadataListe vedlegg, Properties bundle) {
        return vedlegg.vedleggListe.stream()
                .filter(v -> v.status.er(LastetOpp))
                .map(v -> "vedlegg." + v.skjema + "." + v.tillegg + ".tittel")
                .distinct()
                .map(bundle::getProperty)
                .map(navn -> new Vedlegg().withTittel(navn))
                .collect(toList());
    }

    private Date tilDate(LocalDateTime innsendtDato) {
        return Date.from(innsendtDato.atZone(ZoneId.systemDefault()).toInstant());
    }

    private Properties getBundle() {
        return navMessageSource.getBundleFor("soknadsosialhjelp", new Locale("nb", "NO"));
    }

    static String lagEttersendelseLenke(String behandlingsId) {
        return lagContextLenke() + "skjema/" + behandlingsId + "/ettersendelse";
    }

    private String lagFortsettSoknadLenke(String behandlingsId) {
        return lagContextLenke() + "skjema/" + behandlingsId + "/0";
    }

    private static String lagContextLenke() {
        String miljo = System.getProperty("environment.name", "");

        String tjenesterPostfix = "";
        if (miljo.contains("t") || miljo.contains("q")) {
            tjenesterPostfix = "-" + miljo;
        }

        return "https://www" + tjenesterPostfix + ".nav.no/sosialhjelp/soknad/";
    }

}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/api/saksoversikt/SaksoversiktMetadataService.kt
package no.nav.sosialhjelp.soknad.api.saksoversikt

import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.EttersendingsSoknad
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.Hoveddokument
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.InnsendtSoknad
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.PabegyntSoknad
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.Part
import no.nav.sbl.soknadsosialhjelp.tjeneste.saksoversikt.Vedlegg
import no.nav.sosialhjelp.soknad.api.LenkeUtils.lagEttersendelseLenke
import no.nav.sosialhjelp.soknad.api.LenkeUtils.lenkeTilPabegyntSoknad
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata
import no.nav.sosialhjelp.soknad.business.domain.SoknadMetadata.VedleggMetadataListe
import no.nav.sosialhjelp.soknad.db.repositories.soknadmetadata.SoknadMetadataRepository
import no.nav.sosialhjelp.soknad.domain.Vedleggstatus
import no.nav.sosialhjelp.soknad.domain.model.kravdialoginformasjon.SoknadType
import no.nav.sosialhjelp.soknad.ettersending.EttersendingService
import no.nav.sosialhjelp.soknad.ettersending.EttersendingService.Companion.ETTERSENDELSE_FRIST_DAGER
import no.nav.sosialhjelp.soknad.innsending.JsonVedleggUtils.isVedleggskravAnnet
import no.nav.sosialhjelp.soknad.tekster.NavMessageSource
import java.time.Clock
import java.time.LocalDateTime
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Date
import java.util.Locale
import java.util.Properties

class SaksoversiktMetadataService(
    private val soknadMetadataRepository: SoknadMetadataRepository,
    private val ettersendingService: EttersendingService,
    private val navMessageSource: NavMessageSource,
    private val clock: Clock
) {
    fun hentInnsendteSoknaderForFnr(fnr: String): List<InnsendtSoknad> {
        val bundle = bundle
        val soknader = soknadMetadataRepository.hentAlleInnsendteSoknaderForBruker(fnr)
        return soknader.map {
            InnsendtSoknad()
                .withAvsender(
                    Part()
                        .withType(Part.Type.BRUKER)
                        .withVisningsNavn(bundle.getProperty("saksoversikt.mottaker.deg"))
                )
                .withMottaker(
                    Part()
                        .withType(Part.Type.NAV)
                        .withVisningsNavn(bundle.getProperty("saksoversikt.mottaker.nav"))
                )
                .withBehandlingsId(it.behandlingsId)
                .withInnsendtDato(tilDate(it.innsendtDato))
                .withHoveddokument(
                    Hoveddokument()
                        .withTittel(
                            if (it.type == SoknadType.SEND_SOKNAD_KOMMUNAL) {
                                bundle.getProperty("saksoversikt.soknadsnavn")
                            } else {
                                bundle.getProperty("saksoversikt.soknadsnavn.ettersending")
                            }
                        )
                )
                .withVedlegg(tilInnsendteVedlegg(it.vedlegg, bundle))
                .withTema("KOM")
                .withTemanavn(bundle.getProperty("saksoversikt.temanavn"))
                .withLenke(lagEttersendelseLenke(it.behandlingsId))
        }
    }

    fun hentPabegynteSoknaderForBruker(fnr: String): List<PabegyntSoknad> {
        val soknader = soknadMetadataRepository.hentPabegynteSoknaderForBruker(fnr)
        return soknader.map {
            PabegyntSoknad()
                .withBehandlingsId(it.behandlingsId)
                .withTittel("Søknad om økonomisk sosialhjelp")
                .withSisteEndring(tilDate(it.sistEndretDato))
                .withLenke(lenkeTilPabegyntSoknad(it.behandlingsId))
        }
    }

    fun hentSoknaderBrukerKanEttersendePa(fnr: String): List<EttersendingsSoknad> {
        val bundle = bundle
        val ettersendelseFrist = LocalDateTime.now(clock).minusDays(ETTERSENDELSE_FRIST_DAGER.toLong())
        val datoFormatter = DateTimeFormatter.ofPattern("d. MMMM yyyy")
        val soknader = soknadMetadataRepository.hentInnsendteSoknaderForBrukerEtterTidspunkt(fnr, ettersendelseFrist)
        return soknader.map {
            EttersendingsSoknad()
                .withBehandlingsId(it.behandlingsId)
                .withTittel(bundle.getProperty("saksoversikt.soknadsnavn") + " (" + it.innsendtDato.format(datoFormatter) + ")")
                .withLenke(lagEttersendelseLenke(it.behandlingsId))
                .withVedlegg(finnManglendeVedlegg(it, bundle))
        }
    }

    private fun finnManglendeVedlegg(soknad: SoknadMetadata, bundle: Properties): List<Vedlegg> {
        val nyesteSoknad = ettersendingService.hentNyesteSoknadIKjede(soknad)
        return nyesteSoknad.vedlegg.vedleggListe
            .asSequence()
            .filter { it.status.er(Vedleggstatus.VedleggKreves) }
            .filter { !isVedleggskravAnnet(it) }
            .map { "vedlegg." + it.skjema + "." + it.tillegg + ".tittel" }
            .distinct()
            .map { bundle.getProperty(it) }
            .map { Vedlegg().withTittel(it) }
            .toList()
    }

    private fun tilInnsendteVedlegg(vedlegg: VedleggMetadataListe, bundle: Properties): List<Vedlegg> {
        return vedlegg.vedleggListe
            .asSequence()
            .filter { it.status.er(Vedleggstatus.LastetOpp) }
            .map { "vedlegg." + it.skjema + "." + it.tillegg + ".tittel" }
            .distinct()
            .map { bundle.getProperty(it) }
            .map { Vedlegg().withTittel(it) }
            .toList()
    }

    private fun tilDate(innsendtDato: LocalDateTime): Date {
        return Date.from(innsendtDato.atZone(ZoneId.systemDefault()).toInstant())
    }

    private val bundle: Properties get() = navMessageSource.getBundleFor("soknadsosialhjelp", Locale("nb", "NO"))
}
